tests/csapi/apidoc_room_history_visibility_test.go

```diff
@@ -0,0 +1,337 @@
+package csapi_tests
+
+import (
+	"net/http"
+	"testing"
+
+	"github.com/tidwall/gjson"
+
+	"github.com/matrix-org/complement/internal/b"
+	"github.com/matrix-org/complement/internal/client"
+	"github.com/matrix-org/complement/internal/match"
+	"github.com/matrix-org/complement/internal/must"
+	"github.com/matrix-org/complement/runtime"
+)
+
+// TODO most of this can be refactored into data-driven tests
```

Comment:
Up to you if you want to try this out in this PR or not. Idiomatic Go would have table-based tests like:
```go
testCases := []struct{
    visibility string  // shared, joined, etc 
    requesterMembershipState string // typically join, sometimes invite
    wantResponse match.HTTPResponse
    skipOnRuntimes []string // dendrite
}{
    { // example for TestFetchEvent
        visibility: "shared",
        requesterMembershipState: "join",
        wantResponse: match.HTTPResponse{
		StatusCode: 200,
		JSON: []match.JSON{
			// No harm in checking if the event data is also as expected
			match.JSONKeyEqual("content", map[string]interface{}{
				"msgtype": "m.text",
				"body":    "Hello world",
			}),
			match.JSONKeyEqual("type", "m.room.message"),

			// the spec technically doesn't list these following keys, but we're still checking them because sytest did.
			// see: https://github.com/matrix-org/matrix-doc/issues/3540
			match.JSONKeyEqual("room_id", roomID),
			match.JSONKeyEqual("sender", alice.UserID),
			match.JSONKeyEqual("event_id", eventID),
			match.JSONKeyTypeEqual("origin_server_ts", gjson.Number),
		},
	},
    }
}
deployment := Deploy(t, b.BlueprintOneToOneRoom)
defer deployment.Destroy(t)
alice := deployment.Client(t, "hs1", "@alice:hs1")
bob := deployment.Client(t, "hs1", "@bob:hs1")

for _, tc := range testCases {
	roomID := createRoomWithVisibility(t, alice, tc.visibility)
	switch (tc.requesterMembershipState) {
		case "join":
			bob.JoinRoom(t, roomID, nil)
			bob.SyncUntilTimelineHas(t, roomID, func(event gjson.Result) bool {
				return event.Get("type").Str == "m.room.member" &&
					event.Get("content.membership").Str == "join" &&
					event.Get("state_key").Str == bob.UserID
			})
		// etc...
	}

	eventID := alice.SendEventSynced(t, roomID, b.Event{
		Type: "m.room.message",
		Content: map[string]interface{}{
			"msgtype": "m.text",
			"body":    "Hello world",
		},
	})

	res := fetchEvent(t, bob, roomID, eventID)

	must.MatchResponse(t, res, tc.wantResponse)
}
```   

Probably should be wrapped in sub-tests but you get the idea. It would be nice to see this done in this PR given the sheer number of visibility tests this is adding.

Reply:
Hmmm, thanks for the example, I'll try this then, and wrap them in sub-tests